### PR TITLE
sail-riscv-rv32,sail-riscv-rv64: fix build on darwin

### PR DESCRIPTION
**Pull request by Et7f3**
###### Description of changes

Found while testing https://github.com/NixOS/nixpkgs/pull/225044

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Opened-PR upstream and opened it here for visibility.

### CODE REVIEW

File: pkgs/applications/virtualization/sail-riscv/default.nix [hunk 1]
```diff
@@ -24,14 +25,23 @@ stdenv.mkDerivation rec {
   buildInputs = with ocamlPackages; [ zlib linksem ];
   strictDeps = true;
 
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/riscv/sail-riscv/pull/250/commits/8bd37c484b83a8ce89c8bb7a001b8ae34dc4d77f.patch";
```
Comment (Et7f3) [inline pkgs/applications/virtualization/sail-riscv/default.nix:30]: Does this patch handle aarch64-darwin ?

**Comment by GenericNerdyUsername** [comment]
This builds find on x86_64-linux, is there anything else I can do to help?

**Comment by Et7f3** [comment]
Just wait for a @NixOS/darwin-maintainers with aarch64 to test. If fine it can be merged then.

**Comment by StepBroBD** [comment]
On MacOS 13.4 Beta (22F5027f) (`aarch64-darwin`):

```shell
$ nix build .#sail-riscv-rv32 .#sail-riscv-rv64                                                                                                                                                                                                                                                                                                                                                                                                                                                          
error: builder for '/nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv' failed with exit code 1;
       last 10 log lines:
       > no configure script, doing nothing
       > building
       > (cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o ast.lem -picky_multiple_parses true ../../language/sail.ott)
       > Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
       > (cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o jib.lem -picky_multiple_parses true ../../language/jib.ott)
       > Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
       > installing
       > Installing /nix/store/q1yj9z261272zr6qpgxmp10q720hi072-ocaml4.14.1-sail-0.15/lib/ocaml/4.14.1/site-lib/libsail/Libsail_sites.ml
       > Error: Program codesign not found in the tree or in PATH
       > Hint: codesign should be part of the macOS installation
       For full logs, run 'nix log /nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv'.
error: 1 dependencies of derivation '/nix/store/02irmylk1q44i9swz11r44vj9w3bhc40-sail-riscv-0.5.drv' failed to build
```

```shell
$ nix log /nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv                                                                                                                                                                                                                                                                                                                                                                                                                          
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/qh1y634sbqzi1jlv8rdvijy9m1ga5xgn-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Done: 47% (11/23, 12 left) (jobs: 0)^M                                    ^MDone: 47% (11/23, 12 left) (jobs: 0)^M                                    ^MDone: 13% (65/492, 427 left) (jobs: 0)^M                                      ^MDone: 13% (65/492, 427 left) (jobs: 0)^M                                      ^MDone: 37% (249/668, 419 left) (jobs: 1)^M                                       ^MDone: 37% (249/668, 419 left) (jobs: 1)^M                                       ^MDone: 34% (286/>
Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
Done: 62% (764/1226, 462 left) (jobs: 4)^M                                        ^MDone: 62% (771/1229, 458 left) (jobs: 4)^M                                        ^MDone: 62% (771/1229, 458 left) (jobs: 4)^M                                        ^MDone: 62% (780/1243, 463 left) (jobs: 4)^M                                        ^MDone: 62% (780/1243, 463 left) (jobs: 4)^M                                        ^M(cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj823>
Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
Done: 62% (780/1243, 463 left) (jobs: 4)^M                                        ^MDone: 62% (783/1243, 460 left) (jobs: 2)^M                                        ^MDone: 62% (783/1243, 460 left) (jobs: 2)^M                                        ^MDone: 62% (783/1243, 460 left) (jobs: 2)^M                                        ^MDone: 62% (783/1243, 460 left) (jobs: 2)^M                                        ^MDone: 62% (783/1243, 460 left) (jobs: 2)^M                             >
installing
Installing /nix/store/q1yj9z261272zr6qpgxmp10q720hi072-ocaml4.14.1-sail-0.15/lib/ocaml/4.14.1/site-lib/libsail/Libsail_sites.ml
Error: Program codesign not found in the tree or in PATH
Hint: codesign should be part of the macOS installation
```


<details>
<summary>Result of `nix-build -A sail-riscv-rv32`</summary>

```shell
$ nix-build -A sail-riscv-rv32                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
these 2 derivations will be built:
  /nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv
  /nix/store/lmfbdzvqfy6wyf0g5hq2lmk1a0fi0h2a-sail-riscv-0.5.drv
building '/nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv'...
unpacking sources
unpacking source archive /nix/store/qh1y634sbqzi1jlv8rdvijy9m1ga5xgn-source
source root is source
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
no configure script, doing nothing
building
(cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o ast.lem -picky_multiple_parses true ../../language/sail.ott)
Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
(cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o jib.lem -picky_multiple_parses true ../../language/jib.ott)
Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
installing
Installing /nix/store/q1yj9z261272zr6qpgxmp10q720hi072-ocaml4.14.1-sail-0.15/lib/ocaml/4.14.1/site-lib/libsail/Libsail_sites.ml
Error: Program codesign not found in the tree or in PATH
Hint: codesign should be part of the macOS installation
error: builder for '/nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv' failed with exit code 1;
       last 10 log lines:
       > no configure script, doing nothing
       > building
       > (cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o ast.lem -picky_multiple_parses true ../../language/sail.ott)
       > Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
       > (cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o jib.lem -picky_multiple_parses true ../../language/jib.ott)
       > Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
       > installing
       > Installing /nix/store/q1yj9z261272zr6qpgxmp10q720hi072-ocaml4.14.1-sail-0.15/lib/ocaml/4.14.1/site-lib/libsail/Libsail_sites.ml
       > Error: Program codesign not found in the tree or in PATH
       > Hint: codesign should be part of the macOS installation
       For full logs, run 'nix log /nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv'.
error: 1 dependencies of derivation '/nix/store/lmfbdzvqfy6wyf0g5hq2lmk1a0fi0h2a-sail-riscv-0.5.drv' failed to build
```
</details>


<details>
<summary>Result of `nix-build -A sail-riscv-rv64`</summary>

```shell
$ nix-build -A sail-riscv-rv64                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
these 2 derivations will be built:
  /nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv
  /nix/store/02irmylk1q44i9swz11r44vj9w3bhc40-sail-riscv-0.5.drv
building '/nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv'...
unpacking sources
unpacking source archive /nix/store/qh1y634sbqzi1jlv8rdvijy9m1ga5xgn-source
source root is source
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
no configure script, doing nothing
building
(cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o ast.lem -picky_multiple_parses true ../../language/sail.ott)
Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
(cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o jib.lem -picky_multiple_parses true ../../language/jib.ott)
Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
installing
Installing /nix/store/q1yj9z261272zr6qpgxmp10q720hi072-ocaml4.14.1-sail-0.15/lib/ocaml/4.14.1/site-lib/libsail/Libsail_sites.ml
Error: Program codesign not found in the tree or in PATH
Hint: codesign should be part of the macOS installation
error: builder for '/nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv' failed with exit code 1;
       last 10 log lines:
       > no configure script, doing nothing
       > building
       > (cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o ast.lem -picky_multiple_parses true ../../language/sail.ott)
       > Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
       > (cd _build/default/src/lib && /nix/store/xcyis01x17vy0czjfcq5lx2j2wj82394-ott-0.33/bin/ott -sort false -generate_aux_rules true -o jib.lem -picky_multiple_parses true ../../language/jib.ott)
       > Ott version 0.33   distribution of Mon 16 Jan 15:32:01 GMT 2023
       > installing
       > Installing /nix/store/q1yj9z261272zr6qpgxmp10q720hi072-ocaml4.14.1-sail-0.15/lib/ocaml/4.14.1/site-lib/libsail/Libsail_sites.ml
       > Error: Program codesign not found in the tree or in PATH
       > Hint: codesign should be part of the macOS installation
       For full logs, run 'nix log /nix/store/bzafvlp8649ibgxgzw7qvyjprc13j9yz-ocaml4.14.1-sail-0.15.drv'.
error: 1 dependencies of derivation '/nix/store/02irmylk1q44i9swz11r44vj9w3bhc40-sail-riscv-0.5.drv' failed to build
```
</details>

**Comment by reckenrode** [comment]
I tried to add `darwin.sigtool` to the `nativeBuildInputs` for `sail`, but it fails due to trying to sign `.ml` files. Apple’s version of `codesign` bundled with the OS can do that, but `sigtool` only supports signing mach-o binaries.

**Comment by Et7f3** [comment]
I don't have access to aarch64 so can't help debugging. I just marked it broken so it can be merged another maintainer could fix it.

**Pull request merged**